ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52130 [wip] ns for fx: add support for subgraph matching
* #52092 NS for FX: add test for a simple sparsenn model
* **#51669 ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)**
* #51588 Early version of fx graph matcher for NS

Summary:

Adds the basic functionality for the three Numeric Suite core APIs to work on FX models:
1. comparing weights
2. comparing activations, with same input fed to both models
3. comparing activations, with nodes of A shadowing nodes of B

Note: there are a lot of TODOs in the code, and some/most of the APIs and implementation details may change as we iterate.  This is just the first PR.

Test Plan:

We have unit test coverage for all of the APIs, for now this is with toy models:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403094](https://our.internmc.facebook.com/intern/diff/D26403094)

vkuzo committed:
[not ready for review] ns for fx - compare_weights

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - compare_weights"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "[not ready for review] ns for fx - stubs of the three APIs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

hx89 commented on torch/quantization/ns/numeric_suite_core_apis_fx.py:
```diff
@@ -0,0 +1,241 @@
+    2. Lists of Tensors are returned instead of individual Tensors, to unify
+       the return type for calibrating with 1 input vs N inputs.
+    3. `logger_cls` is included in the API for easy result extraction
+    """
```
Just curious why not use get_matching_node_pairs() to find matching activations?

vkuzo replied:
yes, that would work too, it would give the same result

hx89 commented on torch/quantization/ns/numeric_suite_core_apis_fx.py:
```diff
@@ -0,0 +1,241 @@
+                )
+            )
+            if is_logger:
+                results[mod.node_name + '.stats'][mod.model_name] = mod.stats
```
Is it possible that the node_name of the matching nodes are different in model a and b?

vkuzo replied:
Currently, the logger nodes for A and B are inserted by the function which knows about both A and B, so we enforce that the names match when creating the loggers.

If we move to a design where the prepare step only knows about one model, that would no longer work .

vkuzo commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -0,0 +1,346 @@
+    # create the logger object
+    logger_obj = logger_cls(node.name, model_name, other_node_name)
+    # attach the logger object to the parent module
+    setattr(gm, logger_node_name, logger_obj)
```
yes.  This function creates the new node, and in the caller there is logic which links this new node to the next node.  It's done by storing the new node in `env`, and then loading this new node as the argument for a future node.

vkuzo committed:
Update on "ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)"


Summary:

Adds the basic functionality for the three Numeric Suite core APIs to work on FX models:
1. comparing weights
2. comparing activations, with same input fed to both models
3. comparing activations, with nodes of A shadowing nodes of B

Note: there are a lot of TODOs in the code, and some/most of the APIs and implementation details may change as we iterate.  This is just the first PR.

Test Plan:

We have unit test coverage for all of the APIs, for now this is with toy models:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)"


Summary:

Adds the basic functionality for the three Numeric Suite core APIs to work on FX models:
1. comparing weights
2. comparing activations, with same input fed to both models
3. comparing activations, with nodes of A shadowing nodes of B

Note: there are a lot of TODOs in the code, and some/most of the APIs and implementation details may change as we iterate.  This is just the first PR.

Test Plan:

We have unit test coverage for all of the APIs, for now this is with toy models:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)"


Summary:

Adds the basic functionality for the three Numeric Suite core APIs to work on FX models:
1. comparing weights
2. comparing activations, with same input fed to both models
3. comparing activations, with nodes of A shadowing nodes of B

Note: there are a lot of TODOs in the code, and some/most of the APIs and implementation details may change as we iterate.  This is just the first PR.

Test Plan:

We have unit test coverage for all of the APIs, for now this is with toy models:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)"


Summary:

Adds the basic functionality for the three Numeric Suite core APIs to work on FX models:
1. comparing weights
2. comparing activations, with same input fed to both models
3. comparing activations, with nodes of A shadowing nodes of B

Note: there are a lot of TODOs in the code, and some/most of the APIs and implementation details may change as we iterate.  This is just the first PR.

Test Plan:

We have unit test coverage for all of the APIs, for now this is with toy models:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403094](https://our.internmc.facebook.com/intern/diff/D26403094)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx - stubs of the three APIs (compare weights, activations, activations with shadow)"


Summary:

Adds the basic functionality for the three Numeric Suite core APIs to work on FX models:
1. comparing weights
2. comparing activations, with same input fed to both models
3. comparing activations, with nodes of A shadowing nodes of B

Note: there are a lot of TODOs in the code, and some/most of the APIs and implementation details may change as we iterate.  This is just the first PR.

Test Plan:

We have unit test coverage for all of the APIs, for now this is with toy models:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403094](https://our.internmc.facebook.com/intern/diff/D26403094)

[ghstack-poisoned]
